### PR TITLE
Compile speed optimizations

### PR DESCRIPTION
**Pull request by odersky**
Here's the combined effort of Greg and myself this week to make the compiler faster. Overall we seem to be getting between 10 and 15%. We still need to do analysis to find out what optimizations were most beneficial. 

Some optimizations are beneficial for cold compilers and compilers warming up (this is what we see in the Scala build for instance). In particular method splitting seems to help. Other optimizations seem to mostly help in stable state when the compiler is fully jitted. These are harder to predict and often not intuitive at all. For instance method specialization might actually slow down the speed because it makes callsites polymorphic.

Review by @paulp @gkossakowski

### CODE REVIEW

**Comment by paulp** [comment]
Great job on the commit organization.  I know what a pain it can be to maintain it.

**Comment by xeno-by** [comment]
Does not merge cleanly

**Comment by xeno-by** [comment]
Sorry couldn't rebase - there are some genuine conflicts in GenICode. Other than that there are ~5 conflicts with my cleanups, but those can be easily fixed.

File: src/compiler/scala/tools/nsc/ast/Trees.scala [hunk 1]
```diff
@@ -50,7 +50,7 @@ trait Trees extends reflect.internal.Trees { self: Global =>
 
  /** Array selection <qualifier> . <name> only used during erasure */
   case class SelectFromArray(qualifier: Tree, name: Name, erasure: Type)
-       extends TermTree with RefTree
+       extends RefTree with TermTree
```
Comment (jsuereth) [inline src/compiler/scala/tools/nsc/ast/Trees.scala:53]: intuitive?  I'd love to hear what this was.

Comment (paulp) [inline src/compiler/scala/tools/nsc/ast/Trees.scala:53]: I wondered that too, but they made RefTree a class and the class has to come first.

**Comment by odersky** [comment]
Closing this one and submitting another after rebase

**Pull request closed without merging**